yarp/yarp_compiler.c

```diff
@@ -516,6 +516,58 @@ yp_compile_class_path(LINK_ANCHOR *const ret, rb_iseq_t *iseq, const yp_node_t *
     }
 }
 
+/**
+ * In order to properly compile multiple-assignment, some preprocessing needs to
+ * be performed in the case of call or constant path targets. This is when they
+ * are read, the "parent" of each of these nodes should only be read once (the
+ * receiver in the case of a call, the parent constant in the case of a constant
+ * path).
+ */
+static void
+yp_compile_multi_write_lhs(rb_iseq_t *iseq, int lineno, const yp_node_t *node, LINK_ANCHOR *const ret, yp_compile_context_t *compile_context, uint8_t *pushed, bool nested) {
+    NODE dummy_line_node = generate_dummy_line_node(lineno, lineno);
+
+    switch (YP_NODE_TYPE(node)) {
+        case YP_MULTI_TARGET_NODE: {
+            yp_multi_target_node_t *cast = (yp_multi_target_node_t *) node;
+            for (size_t index = 0; index < cast->targets.size; index++) {
+                yp_compile_multi_write_lhs(iseq, lineno, cast->targets.nodes[index], ret, compile_context, pushed, false);
+            }
+            break;
+        }
+        case YP_CONSTANT_PATH_TARGET_NODE: {
+            yp_constant_path_target_node_t *cast = (yp_constant_path_target_node_t *)node;
+            if (cast->parent) {
+                ADD_INSN(ret, &dummy_line_node, putnil);
+                yp_compile_multi_write_lhs(iseq, lineno, cast->parent, ret, compile_context, pushed, false);
+            } else {
+                ADD_INSN1(ret, &dummy_line_node, putobject, rb_cObject);
+            }
+            break;
+        }
+        case YP_CONSTANT_PATH_NODE: {
+            yp_constant_path_node_t *cast = (yp_constant_path_node_t *) node;
+            if (cast->parent) {
+                yp_compile_multi_write_lhs(iseq, lineno, cast->parent, ret, compile_context, pushed, false);
+            } else {
+                ADD_INSN(ret, &dummy_line_node, pop);
+                ADD_INSN1(ret, &dummy_line_node, putobject, rb_cObject);
+            }
+            yp_compile_multi_write_lhs(iseq, lineno, cast->child, ret, compile_context, pushed, cast->parent);
+            break;
+        }
+        case YP_CONSTANT_READ_NODE: {
+            yp_constant_read_node_t *cast = (yp_constant_read_node_t *) node;
+            ADD_INSN1(ret, &dummy_line_node, putobject, nested ? Qfalse : Qtrue);
```

Comment:
Not strictly necessary, but could use `RBOOL` here:

`ADD_INSN1(ret, &dummy_line_node, putobject, RBOOL(!nested));